src/DocumentFormat.OpenXml/Framework/Schema/CompiledParticle.cs

```diff
@@ -0,0 +1,79 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using DocumentFormat.OpenXml.Framework.Schema;
+using DocumentFormat.OpenXml.Validation.Schema;
+using System;
+using System.Collections.Generic;
+using System.Threading;
+
+namespace DocumentFormat.OpenXml.Framework
+{
+    internal class CompiledParticle : IComparer<OpenXmlElement>
+    {
+        private LookupItem[] _lookup;
+
+        public CompiledParticle(ParticleConstraint particle)
+        {
+            Particle = particle;
+        }
+
+        public ReadOnlyArray<LookupItem> Lookup
+        {
+            get
+            {
+                if (_lookup is null)
+                {
+                    Interlocked.Exchange(ref _lookup, ParticleCompiler.Compile(Particle));
```

Comment:
I went back and forth because of the capture the lambda does. Since everywhere else uses Lazy, I'll use that again and do some perf tests later to see if it matters.